### PR TITLE
(#1430) MkHook.json

### PR DESCRIPTION
**Pull request by llorllale**
This PR is for #1430:
* Implemented `MkHook.json()`
* Enabled some tests
* Left puzzle for adding more tests + fixing `MkHooks.create()`

The mock hook's data will be stored as an XML that mirrors [github's json structure](https://developer.github.com/v3/repos/hooks/#get-single-hook).

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1434](https://github.com/jcabi/jcabi-github/issues/1434) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1434](https://github.com/jcabi/jcabi-github/issues/1434) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/C9R53K5JA/2532133699); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/test/java/com/jcabi/github/mock/MkHookTest.java [hunk 1]
```diff
@@ -51,6 +54,11 @@
      */
     private transient Hook hook;
 
+    /**
+     * XML encapsulated by the hook.
+     */
+    private transient XML xml;
```
Comment (paulodamaso) [inline src/test/java/com/jcabi/github/mock/MkHookTest.java]: @llorllale I don't think we should use this xml, it breaks encapsulation; we shouldn't access this directly, see comments below.

File: src/test/java/com/jcabi/github/mock/MkHookTest.java [hunk 2]
```diff
@@ -107,26 +139,28 @@ public void createWithCorrectRepo() throws Exception {
...
      */
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void createWithCorrectId() throws Exception {
         MatcherAssert.assertThat(
             "Hook json returned wrong id",
             this.hook.json().getInt("id"),
-            new IsEqual<Integer>(Tv.FIVE)
+            new IsEqual<Integer>(
+                Integer.parseInt(this.xml.xpath("id/text()").get(0))
```
Comment (paulodamaso) [inline src/test/java/com/jcabi/github/mock/MkHookTest.java]: @llorllale I think that we should test real values here; we are testing if we can get `id` correctly from json, and albeit this implementation works, I think that we should use the real value here instead looking for it in the xml

File: src/test/java/com/jcabi/github/mock/MkHookTest.java [hunk 3]
```diff
@@ -107,26 +139,28 @@ public void createWithCorrectRepo() throws Exception {
...
      */
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void createWithCorrectUrl() throws Exception {
         MatcherAssert.assertThat(
             "Hook json returned wrong url",
             this.hook.json().getString("url"),
             new IsEqual<String>(
-                "https://api.github.com/repos/jcabi/jcabi-github/hooks/1"
+                this.xml.xpath("url/text()").get(0)
```
Comment (paulodamaso) [inline src/test/java/com/jcabi/github/mock/MkHookTest.java]: @llorllale Again, we should test if it's returning the string we set before, not a value searched from the xml.

File: src/test/java/com/jcabi/github/mock/MkHookTest.java [hunk 4]
```diff
@@ -135,13 +169,13 @@ public void createWithCorrectUrl() throws Exception {
...
      */
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void createWithCorrectTestUrl() throws Exception {
         MatcherAssert.assertThat(
             "Hook json returned wrong test_url",
             this.hook.json().getString("test_url"),
             new IsEqual<String>(
-                "https://api.github.com/repos/jcabi/jcabi-github/hooks/1/test"
+                this.xml.xpath("test_url/text()").get(0)
```
Comment (paulodamaso) [inline src/test/java/com/jcabi/github/mock/MkHookTest.java]: @llorllale Same as before

File: src/test/java/com/jcabi/github/mock/MkHookTest.java [hunk 5]
```diff
@@ -150,13 +184,13 @@ public void createWithCorrectTestUrl() throws Exception {
...
      */
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void createWithCorrectPingUrl() throws Exception {
         MatcherAssert.assertThat(
             "Hook json returned wrong ping_url",
             this.hook.json().getString("ping_url"),
             new IsEqual<String>(
-                "https://api.github.com/repos/jcabi/jcabi-github/hooks/1/pings"
+                this.xml.xpath("ping_url/text()").get(0)
```
Comment (paulodamaso) [inline src/test/java/com/jcabi/github/mock/MkHookTest.java]: @llorllale Same as before

File: src/test/java/com/jcabi/github/mock/MkHookTest.java (hunk 1 above)
Comment (llorllale) [reply src/test/java/com/jcabi/github/mock/MkHookTest.java]: @paulodamaso I don't see this as an issue when breaking encapsulation in a test of the affected class

File: src/test/java/com/jcabi/github/mock/MkHookTest.java (hunk 3 above)
Comment (llorllale) [reply src/test/java/com/jcabi/github/mock/MkHookTest.java]: @paulodamaso what we "set" is an XML with a structure that `MkHook` must know and understand. This XML is given to `MkHook` by `MkHooks`. So, we must test that `MkHook` understands this XML. It doesn't really matter what values we use; what matters is that `MkHook#json()` agrees with what's in the XML.

As an aside: I think this relationship between *Mk<iterable>* (eg `MkHooks`) and *Mk<singular>* (eg `MkHook`) should be inverted. The XML is an implementation detail that only `MkHook` should know about. If that were the case, then I think structuring the tests in the way you suggest would be more applicable.

Comment (paulodamaso) [reply src/test/java/com/jcabi/github/mock/MkHookTest.java]: @llorllale I still think that it's not a good idea to use our tests to "double test" our classes (testing the same thing twice, like `this.hook.json().getString("url")` and `this.xml.xpath("url/text()").get(0)`); actually, we must test if our objects are being created with the correct values, and not if the both method are returning the same value (if they both perform wrongly the test can still pass), so that's why I consider importing that we use known and defined values for testing in this case.

About the relationship between our classes, I think that in this case it's ok. This `iterable-singular` relationship is indeed high-coupled because `MkHook` must know from where to read the data for its operations and `MkHooks` must know from where to fetch and create new `MkHook` objects, so I think there's no big deal in using them that way.

The problem that I see is that we shouldn't be operating over our objects internal XML neither retrieving it as json: both of them kills encapsulation. After all, we have a lot of naked data around  the project because we set values in xml and in json everywhere.

Comment (llorllale) [reply src/test/java/com/jcabi/github/mock/MkHookTest.java]: @paulodamaso I don't agree at all. This is not "double-testing".

I was following the direction you set for these tests: couple them all to a single `MkHook` instance. If I do what you want now, then I'll also have "magic values" inside test methods.

I just browsed through some unit test files and I don't see tests coupled to a single test object, so I think that's the first mistake in these tests for `MkHook`.

I will refactor a bit later on. Once I do, I will have the reference values explicit within each test method. Would that satisfy you?

Comment (paulodamaso) [reply src/test/java/com/jcabi/github/mock/MkHookTest.java]: @llorllale Yes, it will, thanks. About the "magic numbers", I'm not against using them in tests; actually, we must use them for equivalence partitioning testing and boundary-value analysis (in this case, our "partition" and "boundaries" are composed of exactly one value, that is the value we expect from the test).

**Comment by llorllale** [comment]
@paulodamaso check now

**Comment by llorllale** [comment]
@paulodamaso I agree it's repetitive, and it's why I followed your original lead in the first place. 

Extracting `MkStorage` would force me to take care with the numbering of each hook in each test.

I've invested too much time in this task already. If it needs fixing then it can be done later.

**Comment by paulodamaso** [comment]
@llorllale Well, just leave a todo then about all this, so it can be fixed later, ok?

**Comment by llorllale** [comment]
@amihaiemil can you please give a final verdict here?

File: src/main/java/com/jcabi/github/mock/MkHook.java [hunk 6]
```diff
@@ -104,6 +105,37 @@ public int number() {
...
+            .add("ping_url", hook.xpath("ping_url/text()").get(0))
+            .add("name", hook.xpath("name/text()").get(0))
+            .add("events", events)
+            .add(
+                "active",
+                Boolean.parseBoolean(hook.xpath("active/text()").get(0))
+            ).add("config", configs)
+            .add("updated_at", hook.xpath("updated_at/text()").get(0))
+            .add("created_at", hook.xpath("created_at/text()").get(0))
+            .build();
```
Comment (amihaiemil) [inline src/main/java/com/jcabi/github/mock/MkHook.java]: @llorllale I don't like the fact that we cherrypick attributes from the XML and create a JsonObject only with them. I think this method should have one line: 

```java
return new JsonNode(xml).json(
    this.storage.xml().nodes(
        String.format(
            "//repo[@coords='%s']/hooks/hook[id='%s']",
            this.coords, this.num
        )
    ).get(0)
);
```

So, simply parse the XML content into JSON and let ``MkHooks.create(...)`` worry about creating the Hook with the right attributes/contents.

Could you do this?

File: src/test/java/com/jcabi/github/mock/MkHookTest.java [hunk 7]
```diff
@@ -29,64 +29,33 @@
  */
```
Comment (amihaiemil) [inline src/test/java/com/jcabi/github/mock/MkHookTest.java]: @llorllale It seems you basically rewrote all the tests that Paulo wrote [here](https://github.com/jcabi/jcabi-github/pull/1428). Why? :D

File: src/main/java/com/jcabi/github/mock/MkHook.java (hunk 6 above)
Comment (llorllale) [reply src/main/java/com/jcabi/github/mock/MkHook.java]: @amihaiemil I was not aware of/didn't remember `JsonNode`. I agree with you, and also think it would clean up the tests *a lot*.

File: src/test/java/com/jcabi/github/mock/MkHookTest.java (hunk 7 above)
Comment (llorllale) [reply src/test/java/com/jcabi/github/mock/MkHookTest.java]: @amihaiemil to decouple the tests.

Thanks for reminding me I was responsible for the testing structure ending up the way it was (I'm referring to the result of [that](https://github.com/jcabi/jcabi-github/pull/1428) PR): I'm too busy these days and somehow that single `MkHook` instance escaped my attention.

**Comment by llorllale** [comment]
@amihaiemil I am thinking about maintaining the tests decoupled just as I've already done. However, using `JsonNode` inside `MkHook#json()` will cut down repetitive code inside tests a lot. I feel this is the best approach.

Please let me know before I proceed.

**Comment by amihaiemil** [comment]
@llorllale 

>I am thinking about maintaining the tests decoupled just as I've already done. However, using JsonNode inside MkHook#json() will cut down repetitive code inside tests a lot. I feel this is the best approach.

Sure, go ahead :D

**Comment by paulodamaso** [comment]
@llorllale ping, we are getting out of time here

**Comment by llorllale** [comment]
@paulodamaso @amihaiemil please check now

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by amihaiemil** [comment]
@rultor merge

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by 0crat** [comment]
The job [#1434](https://github.com/jcabi/jcabi-github/issues/1434) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)
